net.solarnetwork.node.datum.energymeter.mock/src/net/solarnetwork/node/datum/energymeter/mock/MockEnergyMeterDatumSource.java

```diff
@@ -0,0 +1,211 @@
+package net.solarnetwork.node.datum.energymeter.mock;
+
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.context.MessageSource;
+
+import net.solarnetwork.node.DatumDataSource;
+import net.solarnetwork.node.domain.GeneralNodeACEnergyDatum;
+import net.solarnetwork.node.settings.SettingSpecifier;
+import net.solarnetwork.node.settings.SettingSpecifierProvider;
+import net.solarnetwork.node.settings.support.BasicTextFieldSettingSpecifier;
+import net.solarnetwork.node.support.DatumDataSourceSupport;
+
+/**
+ * Mock plugin to be the source of values for a GeneralNodeACEnergyDatum, the
+ * datum will contain random values based around a settable center deviating by
+ * a settable deviation.
+ * 
+ * <p>
+ * This class implements {@link SettingSpecifierProvider} and
+ * {@link DatumDataSource}
+ * </p>
+ * 
+ * @author robert
+ * @version 1.0
+ */
+
+public class MockEnergyMeterDatumSource extends DatumDataSourceSupport
+		implements DatumDataSource<GeneralNodeACEnergyDatum>, SettingSpecifierProvider {
+
+	// default values
+	private String sourceId = "Mock Energy Meter";
+	private String voltagerms = "230";
+	private String frequency = "50";
+	private String resistance = "10";
+	private String inductance = "10";
+
+	private MessageSource messageSource;
+
+	private Integer apparantPow;
+	private Float phaseVolt;
+	private Float phaseCurr;
+	private Integer watt;
+	private Integer realPow;
+	private Integer reacPow;
+	private Float powfac;
+
+	@Override
+	public Class<? extends GeneralNodeACEnergyDatum> getDatumType() {
+		return GeneralNodeACEnergyDatum.class;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see net.solarnetwork.node.DatumDataSource#readCurrentDatum()
+	 * 
+	 * Returns a {@link GeneralNodeACEnergyDatum} filled with mock data. The
+	 * method uses reflection to find setters in {@link
+	 * GeneralNodeACEnergyDatum} and fill them with a number equal to the
+	 * baseReading +- a random number that goes upto deviation.
+	 * 
+	 * @return A {@link GeneralNodeACEnergyDatum}
+	 * 
+	 */
+	@Override
+	public GeneralNodeACEnergyDatum readCurrentDatum() {
+
+		GeneralNodeACEnergyDatum datum = new GeneralNodeACEnergyDatum();
+		datum.setCreated(new Date());
+		datum.setSourceId(sourceId);
+
+		datum.setFrequency(Float.parseFloat(this.frequency));
+		datum.setVoltage(Float.parseFloat(this.voltagerms));
+		calcVariables();
+		datum.setReactivePower(this.reacPow);
+		datum.setApparentPower(this.apparantPow);
+		datum.setPhaseVoltage(this.phaseVolt);
+		datum.setCurrent(this.phaseCurr);
+		datum.setWatts(this.watt);
+		datum.setRealPower(this.realPow);
+		datum.setPowerFactor(this.powfac);
+		return datum;
```

Comment:
The `wattHourReading` is an important value that should be set here because that value is used by most consumers of meter data. For that it should keep track of the last sample time and calculate as the average power between the last and this sample time * the time difference. That is what that [mock OCPP meter](https://github.com/SolarNetwork/solarnetwork-node/blob/45338407ea7d81ea2a7afdcfa72eefcc10485314/net.solarnetwork.node.ocpp.mock/src/net/solarnetwork/node/ocpp/mock/MockMeterDataSource.java#L119) is doing.

Reply:
I have added a wattHours reading but I am not sure if my calculations are correct.